src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JlinkTask.java

```diff
@@ -391,10 +410,99 @@ private JlinkConfiguration initJlinkConfig() throws BadArgs {
             }
             finder = newModuleFinder(options.modulePath, options.limitMods, roots);
         }
+        // Setup and init actions for JDK linkable runtimes
+        LinkableRuntimesResult result = linkableJDKRuntimesInit(finder, roots);
```

Comment:
I think it's clearer to the readers if inlining the code of `linkablejDKRuntimesInit` here.  Also the `--keep-package-modules` check can be moved after `isLinkFromRuntime` is determined so checking `isLinkFromRuntime` is more explicit.

```suggestion
        boolean isLinkFromRuntime = options.modulePath.isEmpty();
        // In case of custom modules outside the JDK we may
        // have a non-empty module path, which must not include
        // java.base. If it did, we link using packaged modules from that
        // module path. If the module path does not include java.base, we must
        // have a linkable JDK runtime. In that case we take the JDK modules
        // from the run-time image.
        if (finder.find("java.base").isEmpty()) {
            isLinkFromRuntime = true;
            ModuleFinder runtimeImageFinder = ModuleFinder.ofSystem();
            finder = combinedFinders(runtimeImageFinder, finder, options.limitMods, roots);
        }

        // --keep-packaged-modules doesn't make sense as we are not linking
        // from packaged modules to begin with.
        if (isLinkFromRuntime && options.packagedModulesPath != null) {
            throw taskHelper.newBadArgs("err.runtime.link.packaged.mods");
        }
```

Reply:
OK, thanks!